---
title: "Vizualisation per genes"
author: "Jakobi Milan"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---


```{r setup, include=FALSE}
source("~/projects/supergenes/fun/fun_preprocess.R")
source("~/projects/supergenes/fun/fun_plots.R")
print("Loading databases...")

if (!exists("epic")) {
  print("Loading Epic metadata .")
  epic_orig = readRDS("~/projects/supergenes/data/EPIC.hg38.manifest.gencode.v22.rds")
  # epic_orig = readRDS("~/projects/datashare/platforms/EPIC.hg38.manifest.rds")
  epic = as.data.frame(epic_orig)
  rownames(epic) = names(epic_orig)
  head(epic)
  pf_chr_colname = "seqnames"
  pf_pos_colname = "start"
  epic = epic[order(epic[[pf_chr_colname]], epic[[pf_pos_colname]]),]
}



if (!exists("trscr_lusc")) {
  print("Loading LUSC transcription data...")
trscr_lusc = readRDS("~/projects/supergenes/data/tcga_studies/study_TCGA-LUSC_trscr.rds")
}


if (!exists("meth_lusc")) {
  print("Loading LUSC methylation data...")
meth_lusc = readRDS("~/projects/supergenes/data/tcga_studies/study_TCGA-LUSC_meth.rds")
}

if(!exists("penda_superup_deregulated")) {
  print("Loading superexpressed genes list...")
  penda_superup_deregulated = as.data.frame(readxl::read_excel("~/projects/supergenes/data/tables_penda.xlsx"))
  rownames(penda_superup_deregulated) <- penda_superup_deregulated[,1]; penda_superup_deregulated <- penda_superup_deregulated[,-1]
  
  for (i in 1:dim(penda_superup_deregulated)[1]){
    if (is.element(i,c(1:3,178:dim(penda_superup_deregulated)[1])))
      {penda_superup_deregulated[i,1] <- 1}
    else {penda_superup_deregulated[i,1] <- 0}
    
    if (i <= 177)
      {penda_superup_deregulated[i,2] <- 1}
    else {penda_superup_deregulated[i,2] <- 0}
    
  }
  penda_superup_deregulated <- rownames(penda_superup_deregulated[penda_superup_deregulated[,2] == 1, ])
  
}

if(!exists("penda_superdown_deregulated")){
  print("Loading underexpressed genes list...")
  penda_superdown_deregulated = as.data.frame(readxl::read_excel("~/projects/supergenes/data/tables_penda.xlsx",sheet = 2))
  rownames(penda_superdown_deregulated) <- penda_superdown_deregulated[,1]; penda_superdown_deregulated <- penda_superdown_deregulated[,-1]
  
  for (i in 1:dim(penda_superdown_deregulated)[1]){
    if (is.element(i,c(1:123,385:dim(penda_superdown_deregulated)[1]))) 
    {penda_superdown_deregulated[i,1] <- 1}
    else {penda_superdown_deregulated[i,1]<- 0}
    
    if(i <= 384)
    {penda_superdown_deregulated[i,2] <- 1}
    else {penda_superdown_deregulated[i,2]<- 0}
    
  }
  
  penda_superdown_deregulated <- rownames(penda_superdown_deregulated[penda_superdown_deregulated[,2] == 1, ])
}
  
  
if(!exists("penda_superconserved")){
  print("Loading conserved genes list...")
  penda_superconserved = as.data.frame(readxl::read_excel("~/projects/supergenes/data/tables_penda.xlsx",sheet = 3))
  rownames(penda_superconserved) <- penda_superconserved[,1]; penda_superconserved <- penda_superconserved[,-1]
    
    for (i in 1:dim(penda_superconserved)[1]){
      if (is.element(i,c(1:33,80:dim(penda_superconserved)[1])))
      {penda_superconserved[i,1] <- 1}
      else {penda_superconserved[i,1] <- 0}
      
      if(i <= 79)
      {penda_superconserved[i,2] <- 1}
      else {penda_superconserved[i,2] <- 0}
    }
  penda_superconserved <- rownames(penda_superconserved[penda_superconserved[,2] == 1, ])
  
}
features = get_features(penda_superdown_deregulated, trscr_lusc, up_str = 2500, dwn_str = 2500, nb_probe_min = 1)
feat_indexed_probes_bin = get_features_bins(penda_superdown_deregulated,index_pathology = 2,trscr_lusc, up_str = 2500, dwn_str = 2500, nb_probe_min = 1)
```

# Getting features for a given gene list.

Please note that some functions are extracted from the previous vignette (LUSC_vizualisation) and won't be explained again here.

We created previously a features list according to a given genes list, which can be browsed using the following function `catalog(features)`

```{r}
catalog <- function(selected_features = features) {
  available_genes <- list(catalog = "catalog", n_genes = "n_genes")
  
  available_genes[["catalog"]] <- noquote(rownames(selected_features))
  available_genes [["n_genes"]] <- noquote(paste("Genes selected :", length(available_genes[[1]]), sep = " "))
  warning("Note that quotes were removed for clarity, please select your genes using it for the plotting function.")
  return(available_genes)
}

catalog(features)
```


# Transcription/Methylation plot

We can see that we have `r catalog(features)$n_genes` available. We will select arbitrarily PREX2 gene for the following examples chunk.

```{r, warning = FALSE}
plot_selected_gene  = function(selected_gene = selected_gene, expr_data = trscr_lusc$data, meth_data = meth_lusc$data, probes_index = feat_indexed_probes, ...){
  
  
  expr_data = expr_data[,intersect(colnames(expr_data),colnames(meth_data))] #### On ne garde en données que les individus pour lesquels on a méthylation ET 
  selected_expr_values = sort(expr_data[selected_gene,]) ##### On sélectionne les données d'expression du gène sélectionné
  selected_probes = feat_indexed_probes[[selected_gene]] #### On sélectionne les probes se trouvants sur le gène sélectionné
  methvals_of_interest = meth_data[intersect(selected_probes,rownames(meth_data)),order(match(colnames(meth_data),names(selected_expr_values)))]
  ## ci-dessus on prend les données de méthylations émises par nos sondes d'intérêts, dans une df triée selon le niveau d'expression du gène sélectionné par patient.
  
  
  
  
  layout(matrix(1:2,1), respect=TRUE)
  tmp = cbind(selected_expr_values,1:length(selected_expr_values))
  expression_plot = plot(tmp, ylab = paste("Patient index ( ordered by level of expression)"), xlab = paste("Expression level"), main = "Expression/Transcription",xlim=c(0,20))
  
  colors=c("green", "black", "red")
  cols = colorRampPalette(colors)(100)
  
  heatmap = image(methvals_of_interest, Rowv = NA, Colv = NA, axes=FALSE, col=cols, main="Methylation")
  axis(side = 1, at=seq(0,1,1/(nrow(methvals_of_interest)-1)), labels = rownames(methvals_of_interest), tick = FALSE, las=2)
  legend("left", c("hypo", "neutral", "hyper"), xpd = TRUE, pch=15, inset = c(-0.35,-0.25), col=c("green","black","red"),bty="n")
  
  mtext(paste(noquote(selected_gene),"Analysis",sep= " "), outer=TRUE,  cex=2, line=-2)
  
}

plot_selected_gene(selected_gene = "PREX2")
```

This plot allows to vizualise, for a given gene, the behavior of ordered individuals gene expression curve along with the methylation heatmap. It is a tool to easily vizualise an obvious correlation (in the common sense) between correlation and expression.


```{r include=FALSE}
binmap = get_binmap(feat_indexed_probes_bin = feat_indexed_probes_bin, binlist = c("bin1","bin2","bin3","bin4","bin5","bin6"))
means_per_bins_per_genes_per_patient = reduce_rows(meth_lusc$data,binmap,mean,na.rm=T)
```



# Probes coordinates among gene bins

```{r, message=FALSE}
plot_selected_probes <- function(selected_gene = "ALDH3B1",binlist=c("bin1","bin2","bin3","bin4","bin5","bin6"),meth_data = meth_lusc$data){

  
  selected_probes = intersect(feat_indexed_probes[[selected_gene]],rownames(meth_data))
  sub_epic<-feat_indexed_probes_bin[[selected_gene]][[1]]
  TSS <- features[selected_gene,"TSS"]
  
  
  layout(matrix(1:2,1), respect=TRUE)
  
  plot(sub_epic$start,rep(1,length(sub_epic$start)),yaxt="n",xaxt="n",xlim=c(TSS-2500,TSS+2500),ylab ="", xlab = "", main="Probes along the gene (in bp)",pch=16)
  text(sub_epic$start,rep(1.1,length(sub_epic$start)), labels=selected_probes,cex = 0.8,srt = 80)
  bins_coord = c(TSS-2500,TSS-1000,TSS-500,TSS,TSS+500,TSS+1000,TSS+2500)
  sapply(bins_coord, function(bin){abline(v=bin,lty=3)})
  axis(side = 1, at=bins_coord, labels = c("-2500","-1000","-500","TSS","+500","+1000","+2500"), tick = FALSE, las=2)
  
  
  
  
  tmp = means_per_bins_per_genes_per_patient[startsWith(rownames(means_per_bins_per_genes_per_patient),selected_gene),]
  xlabs = rownames(tmp)
  boxplot(t(tmp),xaxt="n",main = "Distribution of mean Mval per bin")
  axis(side = 1, at=1:6, labels = xlabs, tick = FALSE, las=2)
  
  mtext(paste(noquote(selected_gene),"Analysis",sep= " "), outer=TRUE,  cex=2, line=-2)
  
}

plot_selected_probes("PREX2")
```

This panel has for purpose to show distribution of probes along the gene (left), and distribution of average methylation level among bins. It shows where the probes are stacked and where they are not, and to charachterize bins mean methylation values distributions.

