---
title: "BedData Vizualisation - Regions of Interest"
author: "Jakobi Milan"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---





```{r eval=TRUE, echo=FALSE, label="loading databases"}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=TRUE, results="hide", message=FALSE, warning = FALSE)

source("~/projects/supergenes/src/fun_preprocess.R")
source("~/projects/supergenes/src/fun_regions.R")


if (!exists("tcga_lusc_DMR")) {
  require(readr)
  print("Loading LUSC DMR metadata...")
tcga_lusc_DMR <- read_delim("~/projects/supergenes/data/dataAnais/TCGA-LUSC_cancer_DMRs.tsv", ",", escape_double = FALSE, trim_ws = TRUE)
}

if (!exists("platform")){
  require(readr)
platform <- as.data.frame(read.table("~/projects/supergenes/data/dataAnais/hg38_genomic_features.bed"))
colnames(platform) <- c("chromosome","start","end","genomic_feature","gene")
}

if (!exists("trscr_lusc")) {
  print("Loading LUSC transcription data...")
trscr_lusc = readRDS("~/projects/tcga_studies/study_TCGA-LUSC_trscr.rds")
}


if (!exists("epic")) {
  print("Loading Epic metadata .")
  epic_orig = readRDS("~/projects/datashare/platforms/EPIC.hg38.manifest.gencode.v22.rds")
  # epic_orig = readRDS("~/projects/datashare/platforms/EPIC.hg38.manifest.rds")
  epic = as.data.frame(epic_orig)
  rownames(epic) = names(epic_orig)
  head(epic)
  pf_chr_colname = "seqnames"
  pf_pos_colname = "start"
  epic = epic[order(epic[[pf_chr_colname]], epic[[pf_pos_colname]]),]
}

hypo_DMR <- tcga_lusc_DMR[which(tcga_lusc_DMR[,"is.hyper"]=="hypo"),]
hyper_DMR <- tcga_lusc_DMR[which(tcga_lusc_DMR[,"is.hyper"]=="hyper"),]


#chr, start, end, width : hg38 position and width of DMR
#DMR_id : unique ID for that DMR, per cancer
#beta_normal, beta_cancer, delta: mean methylation (difference)
#p.adj : modified DMRcate adjusted p-value
#gc.content : G+C content (between 0 and 1)
#no.cpgs.hm450, no.cpgs.genomic : number of CpGs in HM450 array / in genome
#no.cpgs.expected : number of CpGs expected by chance
#cpg.ratio = no.cpgs.genomic/no.cpgs.expected
#num.changes : number of times methylation difference flips withing DMR (used during filtering, after significance testing)
#feature, f_start, f_end : largest feature which overlaps with this DMR, one of INTRON, P2000 or INTER
#feature_gene_name : if feature is INTRON or P2000, annotated gene it belongs to
#closest_gene_name, closest_distance, gene_start, gene_end : gene name, distance and span whose main TSS is closest to DMR
#contains.CGI : whether an annotated CGI is overlapping with the DMR
#category : our annotated DMR category
#is.hyper : whether hyper or hypomethylated
#no.atac.peaks : how many ATAC-seq peaks were found to overlap with the DMR
#selected : hyper: beta_normal<0.5&(atac_peaks==0), hypo:beta_normal>=0.5&(atac_peaks>0)

targeted_genes <- penda_superup_deregulated
DMR_table<-as.data.frame(tcga_lusc_DMR)
source("~/projects/supergenes/src/fun_plot_regions.R")
```

# Foreword

After a first vizualisation using bins ,based on distance to TSS, to divide genes, we now do use regions of interest to partition regions along genes : exons ( noted CDS), introns, UTR'3, UTR'5, intergenes region and a region of 2kbp upstream gene TSS from all transcripts, noted P2000. We also added Differentially Methylated Regions (DMR) on our vizualisation, based on a slightly modified version of DMRcate.

The databatases of DMRs and regions of interest were provided by BSC from Strasbourg. Since they are only interested on protein-coding genes for now, we had to do so to use their data.

# Data presentation

Data table used and presented in the previous vignette are still used. We are going to focus on super upmethylated genes obtained by the Penda method.

We used here a `DMR_table` giving the 11279 DMRs obtained from TCGA data for all cancers, with 28 features : coordinates, wether the region is hypomethylated or hypermethylated, whether an annotated CGI is overlapping with the DMR...
In addition, is also used a `platform` table giving regions (eg : exons) coordinates indexed by genes, for a total of 19632 features. Thus, we have for 174 of our 177 superup methylated genes, coordinates of different regions along them.

# Data preprocessing

## Indexing probes by regions

We created a function taking as input the list of features of interest, the `platform` giving coordinates for regions, the EPIC database and a prebuilt list of probes indexed by features such as given as output from our `get_features` function. We broadened our previously used window, using an interval of -7500,+7500 bp window around the TSS.


```{r}
features <- get_features(targeted_genes, study = trscr_lusc, up_str = 7500, dwn_str = 7500)

get_features_regions <- function(features_list = features, platform_regions = platform, epimed_database = epic, indexed_probes = feat_indexed_probes ){

regions_name <-unique(as.character(platform[,"genomic_feature"]))

  feat_indexed_probes_regions <- lapply(1:nrow(features_list), function(index){


    gene <- names(feat_indexed_probes[index])
    tmp_probes <- feat_indexed_probes[[gene]]
    epic_tmp <- epic[tmp_probes,c("start","end")]
    sub_epic <- epic_tmp[intersect(rownames(epic_tmp), rownames(meth_lusc$data)),]
    pf_gene <- platform[which(platform[,"gene"]==gene),]
  

    pf_regions = lapply(regions_name, function(region){
      regions<-pf_gene[which(pf_gene[,"genomic_feature"]==region),]
      return(regions)
    })

    names(pf_regions) <- regions_name

  per_regions_type_coordinates <-sapply(pf_regions, function(pf){
    if (nrow(pf)>=1){
      per_regions_coordinates = apply(pf,1,function(region){
        range = c(region["start"],region["end"])
      })
    }
  })

  names(per_regions_type_coordinates) <- regions_name
  
  INTER = vector()
  P2000 = vector()
  UTR5 = vector()
  INTRON = vector()
  CDS = vector()
  UTR3 = vector()

  for (i in 1:nrow(sub_epic)){
    
      if (!is.null(per_regions_type_coordinates$INTER))
        if(sum(apply(per_regions_type_coordinates$INTER,2, function(window){
          sub_epic[i,"start"] > window["start"] & sub_epic[i,"end"] < window["end"] }),na.rm=T)>=1)
        {INTER[i]<-rownames(sub_epic[i,])}
    
    
  
      if (!is.null(per_regions_type_coordinates$P2000))
        if(sum(apply(per_regions_type_coordinates$P2000,2, function(window){
          sub_epic[i,"start"] > window["start"] & sub_epic[i,"end"] < window["end"] }),na.rm=T)>=1)
        {P2000[i]<-rownames(sub_epic[i,])}
  
    

      if (!is.null(per_regions_type_coordinates$UTR5))
        if(sum(apply(per_regions_type_coordinates$UTR5,2, function(window){
          sub_epic[i,"start"] > window["start"] & sub_epic[i,"end"] < window["end"] }),na.rm=T)>=1)
        {UTR5[i]<-rownames(sub_epic[i,])}
  
    

      if (!is.null(per_regions_type_coordinates$INTRON))
        if(sum(apply(per_regions_type_coordinates$INTRON,2, function(window){
          sub_epic[i,"start"] > window["start"] & sub_epic[i,"end"] < window["end"] }),na.rm=T)>=1)
        {INTRON[i]<-rownames(sub_epic[i,])}
    
    
      if (!is.null(per_regions_type_coordinates$CDS))
        if(sum(apply(per_regions_type_coordinates$CDS,2, function(window){
          sub_epic[i,"start"] > window["start"] & sub_epic[i,"end"] < window["end"] }),na.rm=T)>=1) 
      {CDS[i]<-rownames(sub_epic[i,])}
    
      if (!is.null(per_regions_type_coordinates$UTR3))
        if(sum(apply(per_regions_type_coordinates$UTR3,2, function(window){
          sub_epic[i,"start"] > window["start"] & sub_epic[i,"end"] < window["end"] }),na.rm=T)>=1)
        {UTR3[i]<-rownames(sub_epic[i,])}
  
    }

  ret <- list(sub_epic = sub_epic, INTER = INTER, P2000 = P2000, UTR5 = UTR5, CDS = CDS, INTRON = INTRON, UTR3 = UTR3)
  return(ret)
  
  })
  names(feat_indexed_probes_regions) <- rownames(features_list)
  return(feat_indexed_probes_regions)

}

feat_indexed_probes_regions = get_features_regions(features_list=features)
```

The obtained output is 2 level depht list, containing for each gene, the probes names for each region of the gene, along with a `sub_epic` list giving all probes indexed for the features and their coordinates. It is notable by exploring the obtained list that probes location are very unbalanced for a given gene : for example, our FKBP4 features has 11 probes on its P2000 region, 1 on exon and 5 on introns, none for the rest.

This note led us to developp a tool to vizualize, for each gene, its region and probe distribution.

## First vizualisation

The following function takes as input a gene of interest, the data table of features such as given by `get_features`, our DMRs coordinates, our regions coordinates given by the `platform` database, the list containings probes indexed by features, and the window around the TSS the user wants to vizualise, which should be the same as selected in the `get_features` function.


```{r}



plot_selected_regions<- function(selected_gene = "FKBP4", features_list = features, DMRtable = DMR_table, DMR_map= DMR_candidates, pf = platform, EPIMEDepic = epic, probes_index = feat_indexed_probes, window=c(7500,7500)){
  
  ## get nearest DMR
  
  DMRs_of_interest<-DMRtable[which(DMRtable[,"closest_gene_name"]==selected_gene),]
  #DMRs_of_interest<- DMR_table[intersect(DMR_table[,"DMR_id"],names(DMR_map[[selected_gene]])),]
  

  
  
  ## get data
    upstream = window[2]
    downstream = window[1]
    pf_gene <- platform[which(platform[,"gene"]==selected_gene),2:4]
    tmp_probes <- feat_indexed_probes[[selected_gene]]
    epic_tmp <- epic[tmp_probes,c("start","end")]
    sub_epic <- epic_tmp[intersect(rownames(epic_tmp), rownames(meth_lusc$data)),]
    TSS <- features_list[selected_gene,"TSS"]
    
    ## get subset of coordinates
    
    introns_coordinates <- pf_gene[which(pf_gene[,"genomic_feature"]=="INTRON"),]
    exons_coordinates <- pf_gene[which(pf_gene[,"genomic_feature"]=="CDS"),]
    utr3_coordinates <- pf_gene[which(pf_gene[,"genomic_feature"]=="3UTR"),]
    utr5_coordinates <- pf_gene[which(pf_gene[,"genomic_feature"]=="5UTR"),]
    intergenic_coordinates <- pf_gene[which(pf_gene[,"genomic_feature"]=="INTER"),]
    p2000_coordinates <- pf_gene[which(pf_gene[,"genomic_feature"]=="P2000"),]
    
    ## plot probes + TSS
    
    plot(sub_epic$start, rep(3,length(sub_epic$start)), pch=3, xlim=c(TSS-downstream,TSS+upstream), cex=1, yaxt="n",
         main = paste0(selected_gene,": Probes repartition among regions (nprobes = ",nrow(sub_epic),")"),
         xlab="Coordinates (in bp)",
         ylab="",
         ylim=c(0,5))
    
    legend("left", c("P2000", "Exons", "Introns","3'UTR","5'UTR","Inter"),inset=c(-0.12,0), xpd = TRUE, pch=15, col=c("grey","blue","red","green","orange","black"),bty="n")
    
    #text(sub_epic$start,rep(4,length(sub_epic$start)), labels = rownames(sub_epic) ,cex = 0.8,srt = 80) ##probes label
    text(TSS,0.2,labels = paste0("TSS : ", TSS),cex = 0.7)
    
    points(TSS, 0.5, pch=9, col="red")
    abline(h=0.5, lty=3, lwd = 1)
    
    ## plot regions
    
    
    
    if(nrow(p2000_coordinates) > 0) {
      apply(p2000_coordinates,1, function(p2000){
        rect(p2000[["start"]],1.5, p2000[["end"]],2.5,
        col="grey",
        border ="black")
      })
    }
    
    
    
    if (nrow(exons_coordinates)>0){
      apply(exons_coordinates,1, function(exon){
        rect(exon[["start"]],1.5, exon[["end"]], 2.5,
               col="blue",
               border="black")
      })
    }
    
    
    if(nrow(introns_coordinates)>0){
      apply(introns_coordinates,1, function(intron){
        rect(intron[["start"]],1.5, intron[["end"]],2.5,
               col="red",
               border="black")
      })
    }
    
    if(nrow(utr3_coordinates)>0){
      apply(utr3_coordinates,1, function(utr3){
        rect(utr3[["start"]],1.5, utr3[["end"]],2.5,
             col="green",
             border="black")
        
      })
    }
    
    
    
    if(nrow(utr5_coordinates)>0){
      apply(utr5_coordinates,1, function(utr5){
        rect(utr5[["start"]],1.5, utr5[["end"]],2.5,
             col="orange",
             border="black")
      })
    }
    
    
    if(nrow(intergenic_coordinates)>0){
      apply(intergenic_coordinates,1, function(inter){
        rect(inter[["start"]],1.5, inter[["end"]],2.5,
             col="black",
             border="black")
      })
    }
    
    
    ##plot DMRs
  
    
    if(nrow(DMRs_of_interest)>0){
        coords <- apply(DMRs_of_interest,1,function(dmr){
            if(dmr[["is.hyper"]]=="hyper"){col="red"}
           else{col="blue"}
            rect(dmr[["start"]],1.1,dmr[["end"]],1.3,
               col=col,
               border=col)
               coord <- as.numeric(c(dmr[["start"]],dmr[["end"]]))
         
        })
      colnames(coords)<-DMRs_of_interest[["DMR_id"]]
      coords <- apply(coords,2,mean)
      text(coords,rep(1.2,length(coords)),names(coords),cex = 0.5)
    }
    

    
    plot_coordinates <<- list(introns_coordinates = introns_coordinates,
               exons_coordinates = exons_coordinates,
               utr3_coordinates = utr3_coordinates,
               utr5_coordinates = utr5_coordinates,
               p2000_coordinates = p2000_coordinates,
               intergenic_coordinates = intergenic_coordinates,
               probes_coordinates = sub_epic,
               DMRs_of_interest = DMRs_of_interest)
    
  
    
  
  
}


plot_selected_regions("OTX1")
plot_selected_regions("MCM2")
plot_selected_regions("TROAP")
```
Crosses stands for probes location (850k BedChip). Red bands stands for hypermethylated DMRs and blue ones for hypomethylated DMRs.

Counter-intuitively, we do not observe the canonical structure of a gene such as P2000-5'UTR-TSS-introns/exons-3'UTR. Plus, we only observe 27 genes of 174 with DMRs in a range of 10kb or less away from the TSS (both sides). There is three patterns of plots : about regular plots, with coherent structure (eg : OTX1), plots with many missings coordinates (eg : MCM2) and genes with chaotic structure.

About DMRs, we re-affected DMRs given by the other team, using the simple following criterium : a DMRs is considered as gene indexed if one of its extremum (its end or its start) is at least than 10kb from the gene's TSS.
We vizualized each plot and decided to exclude the chaotic ones for now, to obtain a proper heatmap of methylation values per type of regions.



```{r}
#mypath <- file.path("/home","jakobim","projects","supergenes","results","Figures","regions_superup",paste("regions", rownames(features)[i], ".jpg", sep = ""))
#zz = c(rep(NA,nrow(features)))
#names(zz)<- rownames(features) 
source("~/projects/supergenes/src/weirdness_script.R")

plots_regions <- list()
for (i in 1:length(rownames(features))){
  jpeg(file=mypath[i])
  plot_selected_regions(rownames(features)[i],window=c(7500,7500))
  plots_regions[[i]] <- recordPlot()

  rm(plot_coordinates)
  graphics.off()
}
names(plots_regions)<-rownames(features)

#replayPlot(plots_regions$FKBP4)
table(weirdness)
```

