---
title: "ROI"
author: "Jakobi Milan"
date: "3 avril 2019"
output: html_document
---

```{r}

source("~/projects/supergenes/src/fun_preprocess.R")


if (!exists("tcga_lusc_DMR")) {
  require(readr)
  print("Loading LUSC DMR metadata...")
tcga_lusc_DMR <- read_delim("~/projects/supergenes/data/dataAnais/TCGA-LUSC_cancer_DMRs.tsv", ",", escape_double = FALSE, trim_ws = TRUE)
}

if (!exists("platform")){
  require(readr)
platform <- as.data.frame(read.table("~/projects/supergenes/data/dataAnais/hg38_genomic_features.bed"))
colnames(platform) <- c("chromosome","start","end","genomic_feature","gene")
}

if (!exists("trscr_lusc")) {
  print("Loading LUSC transcription data...")
trscr_lusc = readRDS("~/projects/tcga_studies/study_TCGA-LUSC_trscr.rds")
}


if (!exists("epic")) {
  print("Loading Epic metadata .")
  epic_orig = readRDS("~/projects/datashare/platforms/EPIC.hg38.manifest.gencode.v22.rds")
  # epic_orig = readRDS("~/projects/datashare/platforms/EPIC.hg38.manifest.rds")
  epic = as.data.frame(epic_orig)
  rownames(epic) = names(epic_orig)
  head(epic)
  pf_chr_colname = "seqnames"
  pf_pos_colname = "start"
  epic = epic[order(epic[[pf_chr_colname]], epic[[pf_pos_colname]]),]
}

hypo_DMR <- tcga_lusc_DMR[which(tcga_lusc_DMR[,"is.hyper"]=="hypo"),]
hyper_DMR <- tcga_lusc_DMR[which(tcga_lusc_DMR[,"is.hyper"]=="hyper"),]


#chr, start, end, width : hg38 position and width of DMR
#DMR_id : unique ID for that DMR, per cancer
#beta_normal, beta_cancer, delta: mean methylation (difference)
#p.adj : modified DMRcate adjusted p-value
#gc.content : G+C content (between 0 and 1)
#no.cpgs.hm450, no.cpgs.genomic : number of CpGs in HM450 array / in genome
#no.cpgs.expected : number of CpGs expected by chance
#cpg.ratio = no.cpgs.genomic/no.cpgs.expected
#num.changes : number of times methylation difference flips withing DMR (used during filtering, after significance testing)
#feature, f_start, f_end : largest feature which overlaps with this DMR, one of INTRON, P2000 or INTER
#feature_gene_name : if feature is INTRON or P2000, annotated gene it belongs to
#closest_gene_name, closest_distance, gene_start, gene_end : gene name, distance and span whose main TSS is closest to DMR
#contains.CGI : whether an annotated CGI is overlapping with the DMR
#category : our annotated DMR category
#is.hyper : whether hyper or hypomethylated
#no.atac.peaks : how many ATAC-seq peaks were found to overlap with the DMR
#selected : hyper: beta_normal<0.5&(atac_peaks==0), hypo:beta_normal>=0.5&(atac_peaks>0)

targeted_genes <- as.character(unique(platform[,5]))
features <- get_features(targeted_genes, study = trscr_lusc)
table(platform[,4])
tmpgene = gene
tmp = features[tmpgene,c("tx_start","tx_end","strand")]
tmp2 = platform[which(platform[,5]==tmpgene),]
tmp2 <-tmp2[order(tmp2[,2]),]
tmp
sub_epic
tmp2
```


```{r}


get_features_regions <- function(targeted_genes, features = features, up_str = 2500, dwn_str = 2500, nb_probe_min = 1, ...) { #### In my uses-case, 1 for LUAD & 2 for LUSC
  print("Creating a list of features...")
  features <- study$platform[intersect(rownames(study$platform), targeted_genes), ]




  print("Indexing probes by features and by bins")

  pf_chr_colname <- "seqnames"
  pf_pos_colname <- "start"
  chrs <- unique(features[, 1])
  chrs_indexed_epic <- lapply(chrs, function(chr) {
    print(chr)
    idx <- rownames(epic)[epic[[pf_chr_colname]] %in% chr]
    ret <- epic[idx, ]
    return(ret)
  })
  names(chrs_indexed_epic) <- chrs




  feat_indexed_probes_regions <- epimedtools::monitored_apply(features, 1, function(gene) {

    chr <- gene[[1]]
    meth_platform <- chrs_indexed_epic[[chr]]
    whole_feature <- dmprocr::get_probe_names(gene, meth_platform, pf_chr_colname, pf_pos_colname, up_str, dwn_str)
    sub_epic <- meth_platform[tmp_probes, c(pf_chr_colname, pf_pos_colname)]
    
    ret <- list(sub_epic = sub_epic, whole_feature = whole_feature, bin1 = bin1, bin2 = bin2, bin3 = bin3, bin4 = bin4, bin5 = bin5, bin6 = bin6)
    return(ret)
    
    
  })



  return(feat_indexed_probes_bin)
}

```


